Yaw in flat target.msg

brunopinto900 opened the pull request:
Problem Description:

This PR adds yaw field to the FlatTarget.msg and any necessary modifications. The circle trajectory now sets yaw to point to the center of the circle at all time.

Additional Context:

This PR is related with #153.

To discuss:

1) The polynomial trajectory getYaw returns 0

```
float polynomialtrajectory::getYaw(double time) {
  float yaw = 0.0;
  return yaw;
}
```

2) Yaw is a float (Float32) not a double (Float64)

Jaeyoung-Lim commented on .gitignore:
```diff
@@ -5,3 +5,6 @@
 .vscode/*
+
+#Rviz config
+geometric_controller/launch/config_file.rviz
```
Why would you want to ignore the rviz config?

brunopinto900 replied:
Because i am always configuring rviz. So need always to commit. I can remove this condition on .gitignore if you like.

Jaeyoung-Lim replied:
Please remove it from the gitignore. You just need to not commit the rviz changes you did when you commit

Jaeyoung-Lim commented on geometric_controller/src/geometric_controller.cpp:
```diff
@@ -147,6 +147,7 @@ void geometricCtrl::flattargetCallback(const controller_msgs::FlatTarget &msg) {
 
   targetPos_ = toEigen(msg.position);
   targetVel_ = toEigen(msg.velocity);
+  mavYaw_ = double(msg.yaw.data);
```
This means that we always need to define yaw, which I think is not what we want. Can you add a case where we can specify ignoring yaw option and handle the case here?

brunopinto900 replied:
You are right. We can solve this by using a mask (IGNORE_YAW) in the message definition. What do you think?

Jaeyoung-Lim replied:
That sounds great

Jaeyoung-Lim commented on trajectory_publisher/src/polynomialtrajectory.cpp:
```diff
@@ -128,6 +128,10 @@ Eigen::VectorXd polynomialtrajectory::getCoefficients(int dim) {
 }
 
+float polynomialtrajectory::getYaw(double time) {
+  float yaw = 0.0;
```
Where does `yaw=0.0` come from? This needs to configurable, and probably belongs to the ignore yaw case

Jaeyoung-Lim reviewed (changes requested):
Thanks for the contribution!

All looks good except I think we need to add the case where yaw can be ignored.

As far as I am concerned this does not add the reduced attitude controller for yaw control. From the discussion we had in https://github.com/Jaeyoung-Lim/mavros_controllers/issues/186 Are you planning to add this?

brunopinto900 commented:
> * I think we need to add the case where yaw can be ignored. Could you add this case to the message definition and handle it in the callback?

What do you mean? Using the already defined mask? something like IGNORE_MASK = some number?

> * The polynomial trajectory is now set to yaw=0. I think this is a unnecessary restriction and basically disables the case for velocity yaw
> * How does `velocity_yaw` option work with this case?

Great question, i still have to think more about this.

> As far as I am concerned this does not add the reduced attitude controller for yaw control. From the discussion we had in #186 Are you planning to add this?

Yes, it doesn't add, i am currently working on it.

Jaeyoung-Lim commented:
> What do you mean? Using the already defined mask? something like IGNORE_MASK = some number?

Yes, preferably a bitmask

> As far as I am concerned this does not add the reduced attitude controller for yaw control. From the discussion we had in #186 Are you planning to add this?
Yes, it doesn't add, i am currently working on it.

Awesome!

Jaeyoung-Lim commented:
@brunopinto900 Any updates?

Jaeyoung-Lim commented:
@brunopinto900 Have you made any progress here?

brunopinto900 commented:
Hey, not yet. I am going on a vacation and have to finnish work stuff.

Jaeyoung-Lim commented:
@brunopinto900 Any updates regarding this work?

Jaeyoung-Lim commented:
@brunopinto900 Do you have an approximate ETA on when this will be addressed?

brunopinto900 commented:
I can finnish tomorrow.

Jaeyoung-Lim commented:
@brunopinto900 Great, thanks

brunopinto900 commented:
I think we need to add the case where yaw can be ignored. Could you add this case to the message definition and handle it in the callback? This means that we always need to define yaw, which I think is not what we want. Can you add a case where we can specify ignoring yaw option and handle the case here?

Jaeyoung-Lim commented:
@brunopinto900 You can define a ignore flag as in https://github.com/mavlink/mavros/blob/29db67693a3fecadfec973a6879930645c2ee9d1/mavros_msgs/msg/AttitudeTarget.msg#L9

brunopinto900 commented:
> * I think we need to add the case where yaw can be ignored. Could you add this case to the message definition and handle it in the callback?

I added "std::msgs/bool IGNORE_YAW" in the message definition? Much simpler than a bit mask.

> * The polynomial trajectory is now set to yaw=0. I think this is a unnecessary restriction and basically disables the case for velocity yaw

The polynomial trajectory now sets yaw to the direction of drone [ atan2f(velocity(1),velocity(0)); ]


> * Please remove it from the gitignore. You just need to not commit the rviz changes you did when you commit

Removed rviz from .gitignore

 ---------------------------------------------------------------------------------------------------

I have two questions. Shouldn't mavYaw_ be targetYaw_? Because its the target orientation and not the current drone yaw

1) In this function:

Eigen::Vector3d geometricCtrl::controlPosition(const Eigen::Vector3d &target_pos, const Eigen::Vector3d &target_vel,
                                               const Eigen::Vector3d &target_acc) {
  /// Compute BodyRate commands using differential flatness
  /// Controller based on Faessler 2017
  const Eigen::Vector3d a_ref = target_acc;
  if (velocity_yaw_) {
    mavYaw_ = getVelocityYaw(mavVel_);
  }

2) mavYaw_ is being computed from the current velocity, shouldn't be from mavVel_? Although in practice, the drone's velocity will closely match the target velocity.

brunopinto900 commented:
Also added the bit mask to ignore yaw. Which one do you prefer, bit mask or bool?

brunopinto900 commented:
Hello @Jaeyoung-Lim Did you the chance to review the pull request?

Jaeyoung-Lim commented:
@brunopinto900 I am closing this PR as stale. Please reopen if you are willing to continue :smile:
